### PR TITLE
impl Convert import to relative path / absolute path (and Convert all... variants) #1589

### PR DESCRIPTION
**Pull request by asukaminato0721**
# Summary

Fixes part of #1589

Added convert‑import code actions (single + all) for relative/absolute paths, wired them into LSP.

# Test Plan

Added tests for both from and import forms.

### CODE REVIEW

**Comment by asukaminato0721** [comment]
The test is missing init parameter, which is in other prs.

**Comment by jvansch1** [comment]
@asukaminato0721 Once merge conflicts are resolved on this PR I can start review.

**Comment by asukaminato0721** [comment]
cc @jvansch1

**Comment by asukaminato0721** [comment]
cc @jvansch1

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    if let Some(action) = convert_all_imports_action(
+        transaction,
+        handle,
+        &module_info,
+        source,
+        ast.as_ref(),
+        ConvertTarget::Absolute,
+    ) {
+        actions.push(action);
+    }
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: nit: can this be a `map` over a list of these possible converttarget so we don't duplicate the code so much?

**Review by kinto0 — CHANGES_REQUESTED** [review]

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+    impl<'a> Visitor<'a> for ImportCollector<'a> {
+        fn visit_stmt(&mut self, stmt: &'a Stmt) {
+            match stmt {
+                Stmt::Import(import) => self.imports.push(ImportStmtRef::Import(import)),
+                Stmt::ImportFrom(import_from) => {
+                    self.imports.push(ImportStmtRef::ImportFrom(import_from));
+                }
+                _ => {}
+            }
+            ruff_python_ast::visitor::walk_stmt(self, stmt);
```
[the remark this answers is not in the document]
Comment (kinto0) [reply pyrefly/lib/state/lsp/quick_fixes/convert_import.rs:334]: what do you think?

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    }
+}
+
+fn split_relative_module(relative_module: &str, absolute_module: &str) -> Option<(String, String)> {
+    if relative_module == "." {
+        let leaf = absolute_module
+            .rsplit_once('.')
+            .map_or(absolute_module, |(_, leaf)| leaf);
+        return Some((".".to_owned(), leaf.to_owned()));
+    }
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: AI comment: wrong dot count for multi-level relative import X. In split_relative_module, rsplit_once('.') miscounts leading dots when the relative module is N≥2 dots directly followed by a leaf:

  import baz            # in foo/bar.py, baz.py at repo root
    relative_module_name_between → "..baz"   (see module_name.rs test line 666)
    split_relative_module("..baz", "baz")
      "..baz".rsplit_once('.') → (".", "baz")   # base=".", non-empty, kept
      → "from . import baz"                       # ❌ should be "from .. import baz"
  from . import baz resolves to foo.baz, which doesn't exist — a semantically broken edit. ...leaf is likewise off by one dot. The .foo.baz case works only because the intermediate package absorbs the miscount. Suggested fix — count leading dots explicitly instead of rsplit_once:
  let dots = relative_module.len() - relative_module.trim_start_matches('.').len();
  let rest = &relative_module[dots..];
  let (pkg, leaf) = rest.rsplit_once('.').unwrap_or(("", rest));
  let base = format!("{}{}", ".".repeat(dots), pkg); // ".." + "" = "..", "." + "foo" = ".foo"

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    if let Some((base, leaf)) = relative_module.rsplit_once('.') {
+        let base = if base.is_empty() { "." } else { base };
+        return Some((base.to_owned(), leaf.to_owned()));
+    }
+    Some((".".to_owned(), relative_module.to_owned()))
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: is this unreachable? can we verify that with the type system or make unreachable!()

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    } else {
+        default_name.to_owned()
+    }
+}
+
+fn relative_module_string(from: &Handle, to: &Handle) -> Option<String> {
+    let module =
+        ModuleName::relative_module_name_between(from.path().as_path(), to.path().as_path())?;
+    let raw = module.as_str();
+    if raw.is_empty() {
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: is this reachable? can we avoid it?

File: pyrefly/lib/test/lsp/code_actions.rs [hunk 6]
```diff
@@ -4987,3 +5010,105 @@ class Derived(Base):
...
+        .position(|title| title == "Convert import to relative path")
+        .unwrap();
+    let updated = apply_refactor_edits_for_module(&module_info, &actions[idx]);
+    let expected = r#"
+# IMPORT-START
+from . import b as mod_b
+# IMPORT-END
+"#;
+    assert_eq!(expected.trim(), updated.trim());
+}
```
Comment (kinto0) [inline pyrefly/lib/test/lsp/code_actions.rs:5954]: Untested edge cases (all reachable through the code paths):
  - Multi-level relative (.., ...) — the buggy case above; not covered by any test.
  - from . import x (module=None, level>0) → absolute. convert_from_import_to_absolute handles it (new_maybe_relative(is_init, level, None) → from pkg import x) but nothing exercises it.
  - handle_require_absolute_import = true (stdlib / search-path / site-packages) should suppress the relative action — logic looks correct, untested.
  - Grouping: import a as x, b as y → merged from . import a as x, b as y. Untested.
  - Multi-line parenthesized from x import (A, B) collapses to a single line (and any comments inside the parens are dropped, since render_imported_names only joins names). Trailing end-of-line comments are safe (outside stmt.range()), but the in-parens case is lossy and untested.
  - is_init files: relative levels differ for __init__.py; no __init__ test.

**Review by kinto0 — CHANGES_REQUESTED** [review]

File: crates/pyrefly_python/src/module_name.rs [hunk 7]
```diff
@@ -389,7 +389,12 @@ impl ModuleName {
                 }
             };
         }
-        Self::from_relative_path_components(components).ok()
+        let module = Self::from_relative_path_components(components).ok()?;
+        if to.file_stem()?.to_str()? == dunder::INIT {
+            Some(Self::from_string(format!(".{}", module.as_str())))
+        } else {
+            Some(module)
```
Comment (kinto0) [inline crates/pyrefly_python/src/module_name.rs:407]: Blocker: the module_name.rs change is wrong for descendant packages

  ModuleName::relative_module_name_between (crates/pyrefly_python/src/module_name.rs:376) is shared — its only other caller is the auto-import path at pyrefly/lib/state/ide.rs:350. The PR appends a dot
  whenever to is an __init__.py. That over-applies.

  The real defect is narrower: from_parts joins with . between components, so 1 + levels_up empty components render as levels_up dots. That only loses a dot when the part below the dots is empty:

  ┌────────────────┬─────────────────────┬─────────┬───────────┬──────────┐
  │      from      │         to          │ correct │ before PR │ after PR │
  ├────────────────┼─────────────────────┼─────────┼───────────┼──────────┤
  │ foo/bar.py     │ foo/__init__.py     │ .       │ ""        │ . ✅     │
  ├────────────────┼─────────────────────┼─────────┼───────────┼──────────┤
  │ foo/bar/baz.py │ foo/__init__.py     │ ..      │ .         │ .. ✅    │
  ├────────────────┼─────────────────────┼─────────┼───────────┼──────────┤
  │ foo/bar.py     │ foo/sub/__init__.py │ .sub    │ .sub      │ ..sub ❌ │
  └────────────────┴─────────────────────┴─────────┴───────────┴──────────┘

  ..sub from foo/bar.py resolves to a sibling of foo. This hits both the new code action (import foo.sub as s → from .. import sub as s) and the pre-existing relative auto-import. Condition on the collapse,
  not on the filename:

  let module = Self::from_relative_path_components(components).ok()?;
  // `join` puts separators *between* components, so an all-dots result is one dot short.
  if module.as_str().chars().all(|c| c == '.') {
      Some(Self::from_string(format!(".{}", module.as_str())))
  } else {
      Some(module)
  }

  Add assert_module_name("foo/bar.py", "foo/sub/__init__.py", ".sub") to test_relative_module_name_between (module_name.rs:657) — no existing case covers a target __init__.py below the source.

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+enum ImportStmtRef<'a> {
+    Import(&'a StmtImport),
+    ImportFrom(&'a StmtImportFrom),
+}
+
+pub(crate) fn convert_import_code_actions(
+    transaction: &Transaction<'_>,
+    handle: &Handle,
+    selection: TextRange,
+) -> Option<Vec<LocalRefactorCodeAction>> {
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: nit: should this be `option<vec1` because an empty list doesn't make sense?

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+
+    if actions.is_empty() {
+        None
+    } else {
+        actions.sort_by(|a, b| a.title.cmp(&b.title));
+        Some(actions)
+    }
+}
+
+fn convert_single_import_action(
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: nit: what's the contract here? when will this return empty?

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+        ConvertTarget::Absolute => "Convert all imports to absolute path",
+    };
+    Some(LocalRefactorCodeAction {
+        title: title.to_owned(),
+        edits,
+        kind: CodeActionKind::REFACTOR_REWRITE,
+    })
+}
+
+fn convert_import_stmt(
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: what's the contract here? when will this return empty?

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+    handle: &Handle,
+    source: &str,
+    stmt: ImportStmtRef<'_>,
+    target: ConvertTarget,
+) -> Option<(TextRange, String)> {
+    let stmt_range = match stmt {
+        ImportStmtRef::Import(import) => import.range(),
+        ImportStmtRef::ImportFrom(import_from) => import_from.range(),
+    };
+    if source
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: is this the best way to do this? do we do something similar in other code actions? why / why not?

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+        _ => return None,
+    };
+    Some((stmt_range, new_text))
+}
+
+fn convert_import_to_relative(
+    transaction: &Transaction<'_>,
+    handle: &Handle,
+    import: &StmtImport,
+) -> Option<Vec<String>> {
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: signature is not clear from looking at this

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+    }
+    Some(
+        grouped
+            .into_iter()
+            .map(|(base, names)| format!("from {base} import {}", names.join(", ")))
+            .collect(),
+    )
+}
+
+fn convert_from_import_to_relative(
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs:200]: can we share more logic with the function above?

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+    handle: &Handle,
+    import_from: &StmtImportFrom,
+) -> Option<String> {
+    let module = handle.module().new_maybe_relative(
+        handle.path().is_init(),
+        import_from.level,
+        import_from.module.as_ref().map(|module| &module.id),
+    )?;
+    let module_str = module.as_str();
+    if module_str.is_empty() {
```
Comment (kinto0) [inline pyrefly/lib/state/lsp/quick_fixes/convert_import.rs:223]: why this guard? when would this happen?

**Review by kinto0 — CHANGES_REQUESTED** [review]
some concerns upon closer look. could you audit the high-level design and make sure API boundaries are clean?

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs (hunk 8 above)
Comment (asukaminato0721) [reply pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: fixed

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs (hunk 13 above)
Comment (asukaminato0721) [reply pyrefly/lib/state/lsp/quick_fixes/convert_import.rs:200]: done

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs (hunk 9 above)
Comment (asukaminato0721) [reply pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: It returns None when the selected import cannot be safely converted to the requested target.

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs (hunk 12 above)
Comment (asukaminato0721) [reply pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: renamed the function to convert_plain_import_to_relative

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs (hunk 14 above)
Comment (asukaminato0721) [reply pyrefly/lib/state/lsp/quick_fixes/convert_import.rs:223]: `from . import x`

File: pyrefly/lib/state/lsp/quick_fixes/convert_import.rs (hunk 10 above)
Comment (asukaminato0721) [reply pyrefly/lib/state/lsp/quick_fixes/convert_import.rs]: when no import in the file has an applicable.
